addons/edi_stock/models/edi_move_report_document.py

```diff
@@ -0,0 +1,144 @@
+        `~._edi_move_report_via` is allowed to be overridden in models
+        inheriting from this one. This method gets called for each derived
+        model and creates the appropriate index for each one.
+        """
+        super(EdiMoveReportDocument, self).init()
+
+        Move = self.env["stock.move"]
+
+        if self._edi_move_report_via is not None:
+            tools.create_index(
+                self._cr,
+                "stock_move_state_picking_type_id_%s_index" % (self._edi_move_report_via,),
+                Move._table,
+                ["state", "picking_type_id", self._edi_move_report_via],
+            )
+        else:
+            tools.create_index(
+                self._cr,
+                "stock_move_state_picking_type_id_index",
+                Move._table,
+                ["state", "picking_type_id"],
+            )
+
+    @api.model
+    def move_report_record_model(self, doc, supermodel="edi.move.report.record"):
+        """Get EDI stock move report record model class
+
+        Subclasses should never need to override this method.
+        """
+        return self.record_model(doc, supermodel=supermodel)
+
+    @api.model
+    def move_report_domain(self, doc):
+        """Get stock move search domain
+
+        The default implementation returns all completed moves
+        associated with the specified picking types for which a report
+        has not yet been generated
+        """
+        domain = [
+            ("picking_type_id", "in", doc.doc_type_id.pick_type_ids.ids),
+            ("state", "=", "done"),
+        ]
+        if self._edi_move_report_via is not None:
+            domain.append((self._edi_move_report_via, "=", False))
+        return domain
+
+    @api.model
+    def move_report_list(self, _doc, moves):
+        """Get list of stock moves for which reports should be generated
+
+        Returns an iterable of ``stock.move`` recordsets.  Each
+        recordset in the iterable will result in the creation of a
+        single EDI stock move report record.
+
+        Note that a recordset is itself an iterable of (singleton)
+        recordsets.  This method may therefore return a single
+        ``stock.move`` recordset, in which case each individual
+        ``stock.move`` record will result in the creation of a
+        separate EDI stock move report record.  The default
+        implementation does exactly this.
+        """
+        return moves
+
+    @api.model
+    def prepare(self, doc):
+        """Prepare document"""
+        MoveReport = self.move_report_record_model(doc)
+        Move = self.env["stock.move"]
+        # Lock moves to prevent concurrent report generation attempts
+        moves = Move.search(self.move_report_domain(doc), order="id")
+        if self._edi_move_report_via is not None:
+            moves.write({self._edi_move_report_via: False})
+        movelist = (x.with_prefetch(moves._prefetch) for x in self.move_report_list(doc, moves))
```

Comment:
Suggestion: Use `mr` instead of `x` as we're dealing with move reports.

Reply:
gonna go with `move` as `self.move_report_list()` returns `stock.move` iterables